sdk/include/opentelemetry/sdk/trace/span_data.h

```diff
@@ -76,6 +86,11 @@ class SpanData final : public Recordable
     parent_span_id_ = parent_span_id;
   }
 
+  void SetAttribute(nostd::string_view key, common::AttributeValue &&value) noexcept override
+  {
+    attributes_[std::string(key)] = value;
```

Comment:
Why can't we use nostd::string_view instead of std::string here for the map?

Since we went with **string_view** in the first place, we were bothered to avoid the unnecessary memcpy. Transforming into string essentially invalidates the optimization done at top-level.

Reply:
See my comment above.

>  Transforming into string essentially invalidates the optimization done at top-level.

Optimized approaches are possible when using own `Recordable` implementations, where there is no need for an intermediate copy is . `SpanData` needs to hold internal copies of values due to lifetime issues.

Reply:
This makes default **SpanData** implementation making assumption that the flow is asynchronous. Whereas there could be a synchronous exporter not needing a memcpy. Another approach is for Exporter to perform serialization on customer thread, thus not needing a copy. Can you clarify on what model should be followed for implementing a synchronous exporter? Does it mean that synchronous exporter devs must implement their own SpanData?

I'm also worried about the unnecessary copies of events added using AddEvent. If we are to maintain a separate container for the Span attributes, how is it going to be done for Events added via AddEvent -- would SpanData provide yet another container for Events too?

Reply:
> This makes default SpanData implementation making assumption that the flow is asynchronous.

Yes, that's an assumption the SpanData implementation is making.

> Can you clarify on what model should be followed for implementing a synchronous exporter? Does it mean that synchronous exporter devs must implement their own SpanData?

They need not, but they surely can. I expect that the vast majority of real-world exporters will come together with custom `Recordable` implementations. OTLP and GCP exporters will.